### ISSUE TITLE
RFE: Automate dev env setup using tox

### ISSUE THREAD

**Issue by mriedem** [jupyterhub/jupyterhub#2961]
This isn't a bug, just a request for enhancement from a new contributor and maybe this has been discussed before in a mailing list thread or meeting (I didn't see an existing issue for this).

I was going through https://github.com/jupyterhub/jupyterhub/blob/master/CONTRIBUTING.md and the dev environment setup seems like it could be pretty easily scripted in a `tox.ini` file and then running tests could be a specific `testenv` where the common setup is handled automatically and the `testenv` just runs `pytest` or a linter or builds docs, whatever.

Basically consider taking the scripts portion of the travis config and putting that into a tox.ini and then you could update the travis config to just run the tox environment for each travis build. Then local dev / test and CI are aligned.

Anyway, just an idea for discussion and like I said, maybe it's been discussed before and if so feel free to close this with a link to whatever discussion was captured (mailing list, meeting notes, forum, etc).

**Comment by mriedem** [comment]
Some related discussion in the forum here:

https://discourse.jupyter.org/t/recommended-way-to-install-jupyterlab-globally-on-mac/2825/2

Recommendations there are to use docker or conda for environment isolation. So there are lots of ways to solve this same problem, though I think being able to use tox locally and in travis CI the same way would be a win so that you know the test results you get locally are the same you're going to get in CI (for the most part), i.e. avoid surprises with travis job failures after you push up a change.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/recommended-way-to-install-jupyterlab-globally-on-mac/2825/7

**Comment by mriedem** [comment]
This is another thing that could be automated with tox:

https://jupyterhub.readthedocs.io/en/stable/contributing/docs.html

So then you could just do something like `tox -e docs` and get the docs built for you without the manual steps in that guide.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/thoughts-on-using-tox/3497/1

**Comment by betatim** [comment]
Closing this as the same comment got posted to the forum and discussion has taken off [there](https://discourse.jupyter.org/t/thoughts-on-using-tox/3497).
